src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -42,301 +42,289 @@
+                    "' for " + Arrays.toString(ctorParameterTypes) + " using lookup " + lookup, e);
         }
     }
 
-    private static <U> BinaryOperator<U> throwingMerger() {
-        return (a, b) -> {
-            throw new IllegalArgumentException("Duplicate keys: " + a);
-        };
+    private MethodHandle methodHandle(RecordComponent component) {
+
+        var pathElement = MemoryLayout.PathElement.groupElement(component.getName());
+        var componentLayout = layout.select(pathElement);
+        var byteOffset = layout.byteOffset(pathElement) + offset;
+        try {
+            return switch (componentLayout) {
+                case ValueLayout vl -> methodHandle(vl, component, byteOffset);
+                case GroupLayout gl -> methodHandle(gl, component, byteOffset);
+                case SequenceLayout sl -> methodHandle(sl, component, byteOffset);
+                case PaddingLayout __ -> throw fail(component, componentLayout);
+            };
+        } catch (NoSuchMethodException | IllegalAccessException e) {
+            throw new InternalError(e);
+        }
+    }
+
+    private MethodHandle methodHandle(ValueLayout vl,
+                                      RecordComponent component,
+                                      long byteOffset) throws NoSuchMethodException, IllegalAccessException {
+
+        assertTypesMatch(component, component.getType(), vl);
+        var mt = MethodType.methodType(vl.carrier(), valueLayoutType(vl), long.class);
+        var mh = LOOKUP.findVirtual(MemorySegment.class, "get", mt);
+        // (MemorySegment, OfX, long) -> (MemorySegment, long)
+        mh = MethodHandles.insertArguments(mh, 1, vl);
+        // (MemorySegment, long) -> (MemorySegment)
+        return castReturnType(MethodHandles.insertArguments(mh, 1, byteOffset), component.getType());
+    }
+
+    private MethodHandle methodHandle(GroupLayout gl,
+                                      RecordComponent component,
+                                      long byteOffset) throws NoSuchMethodException, IllegalAccessException {
+        // Simply return the raw MethodHandle of the recursively computed record mapper
+        return recordMapper(component.getType(), gl, byteOffset).ctor;
+    }
+
+    private MethodHandle methodHandle(SequenceLayout sl,
+                                      RecordComponent component,
+                                      long byteOffset) throws NoSuchMethodException, IllegalAccessException {
+
+        String name = component.getName();
+        var componentType = component.getType();
+        if (!componentType.isArray()) {
+            throw new IllegalArgumentException("Unable to map '" + sl +
+                    "' because the component '" + componentType.getName() + " " + name + "' is not an array");
+        }
+
+        MultidimensionalSequenceLayoutInfo info = MultidimensionalSequenceLayoutInfo.of(sl, componentType);
```

Comment:
Now that we have offsets in the right places, I feel like this can be simplified using recursion (similarly to what I proposed to do in the javadoc). Although here we need to special case element layouts for which a faster option is available. So, something like:

* if element layout is a value layout V, use segment.toArray(V)
* otherwise, retrieve the objects corresponding to the array elements (using the mapper recursively) and return an array of them (it is likely that this part is more imperative, and harder to express with method handles).

The key difference is that the second step above only concerns about going "one dimension deeper".